src/Bicep.Core/Parsing/Parser.cs

```diff
@@ -46,30 +47,54 @@ public override ProgramSyntax Program()
             return programSyntax;
         }
 
-        protected override SyntaxBase Declaration() =>
+        protected override SyntaxBase Declaration(params string[] allowedIdentifiers) =>
```

Comment:
I'd call the parameter `expectingKeywords` since a declaration always starts with a keyword, not an identifier.

Reply:
Ok. I'll refactor this part a bit to reduce number of `when` statements

Reply:
I think a better approach could be to avoid passing this parameter to `Declaration`. In `BaseParser` we can just call `DecorableSyntaxLeadingNodes` (if an `@` token is observed) and `ResourceDeclaration`.

Reply:
True, but I thought that this could be useful in future - i.e. if we would like to have an if block that would allow only resources and modules without outputs or variables - we could re-use it.

 I refactored it to use a local method that would return to switch statement the keyword if it's allowed or null if not valid (to trigger the default case)

Reply:
Yeah, that's a fair point. Btw, I was going to suggest the exact same change you just made to simplify the validation logic for `expectedKeywords` 😄.

Reply:
I got a bit rusty in writing in C#, needed to sleep on it to find a better way ;)